### PR TITLE
Capture values for bailout incrementally across blocks

### PR DESCRIPTION
**Pull request by ThomsonTan**
Capture values for bailout are built incrementally by selecting captured values in one of its predecessor block and mark all not uncommon symbols from all predecessor blocks as recapture when building new bailout.

### CODE REVIEW

File: lib/Backend/InlineeFrameInfo.h [hunk 1]
```diff
@@ -14,6 +14,7 @@ struct BailoutConstantValue {
     };
     void InitVarConstValue(Js::Var value);
     void InitFloatConstValue(FloatConstType value) { this->type = TyFloat64; this->u.floatConst.value = value; }
+    bool operator==(const BailoutConstantValue & bailoutConstValue);
```
Comment (LouisLaf) [inline lib/Backend/InlineeFrameInfo.h]: I don't like operator overloading in this way as I hides the true cost and meaning of the operation.  An IsEqual() function accomplishes the same but I can easy browse to the definition now to see what the code is actually doing.

**Comment by LouisLaf** [comment]
Sorry, there was no @LouisLaf here which is why it took me a while to get to it...

File: lib/Backend/GlobOptBailOut.cpp [hunk 2]
```diff
@@ -951,7 +957,7 @@ GlobOpt::FillBailOutInfo(BasicBlock *block, BailOutInfo * bailOutInfo)
                 {
                     sym = opnd->GetStackSym();
                     Assert(FindValue(sym));
-                    this->blockData.capturedArgs->Set(sym->m_id);
+                    this->blockData.changedSyms->Set(sym->m_id);
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: >sym [](start = 53, length = 3)

Why?  What is changing about sym here?  This at a minimum needs a comment.

Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: Args need to be captured even if without setting, will add a comment.

---
In reply to: [71631271](https://github.com/Microsoft/ChakraCore/pull/1247#discussion_r71631271) [](ancestors = 71631271)

File: lib/Backend/GlobOpt.cpp [hunk 3]
```diff
@@ -1829,6 +1826,58 @@ GlobOpt::CloneValues(BasicBlock *const toBlock, GlobOptBlockData *toData, GlobOp
...
 }
 
+template <typename CapturedList, typename CapturedItemsAreEqual>
+void
+GlobOpt::MergeCapturedValues(
+    GlobOptBlockData * toData,
+    SListBase<CapturedList> * toList,
+    SListBase<CapturedList> * fromList,
+    CapturedItemsAreEqual itemsAreEqual)
+{
```
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: Add comment explaining how this code works and why you can't edit the final list.

**Comment by LouisLaf** [comment]
:shipit:

**Pull request merged**